crypto/src/dex/swap.rs

```diff
@@ -29,40 +29,39 @@ pub static DOMAIN_SEPARATOR: Lazy<Fq> =
 pub struct BatchSwapOutputData {
     pub delta_1: u64,
     pub delta_2: u64,
-    pub lambda_1: u64,
-    pub lambda_2: u64,
+    pub lambda_1_1: u64,
+    pub lambda_2_1: u64,
+    pub lambda_1_2: u64,
+    pub lambda_2_2: u64,
     pub height: u64,
     pub trading_pair: TradingPair,
-    pub success: bool,
 }
 
 impl BatchSwapOutputData {
     /// Given a user's inputs `(delta_1_i, delta_2_i)`, compute their pro rata share
     /// of the batch output `(lambda_1_i, lambda_2_i)`.
     pub fn pro_rata_outputs(&self, (delta_1_i, delta_2_i): (u64, u64)) -> (u64, u64) {
-        if self.success {
-            // The swap succeeded, so the pro rata share is a share of the output amount of
-            // the opposite token type.
-            // The pro rata fraction is delta_j_i / delta_j, which we can multiply through:
-            //   lambda_2_i = (delta_1_i / delta_1) * lambda_2
-            //   lambda_1_i = (delta_2_i / delta_2) * lambda_1
-            // But we want to compute these as
-            //   lambda_2_i = (delta_1_i * lambda_2) / delta_1
-            //   lambda_1_i = (delta_2_i * lambda_1) / delta_2
-            // so that we can do division and rounding at the end.
-            let lambda_2_i = ((delta_1_i as u128) * (self.lambda_2 as u128))
-                .checked_div(self.delta_1 as u128)
+        // The pro rata fraction is delta_j_i / delta_j, which we can multiply through:
+        //   lambda_2_i = (delta_1_i / delta_1) * lambda_2_1 + (delta_2_i / delta_2) * lambda_2_2
+        //   lambda_1_i = (delta_1_i / delta_1) * lambda_1_1 + (delta_2_i / delta_2) * lambda_1_2
+        // But we want to compute these as
+        //   lambda_2_i = (delta_1_i * lambda_2_1) / delta_1 + (delta_2_i * lambda_2_2) / delta_2
+        //   lambda_1_i = (delta_1_i * lambda_1_1) / delta_1 + (delta_2_i * lambda_1_2) / delta_2
+        // so that we can do division and rounding at the end.
+        let lambda_2_i = ((delta_1_i as u128) * (self.lambda_2_1 as u128))
+            .checked_div(self.delta_1 as u128)
+            .unwrap_or(0)
+            + ((delta_2_i as u128) * (self.lambda_2_2 as u128))
+                .checked_div(self.delta_2 as u128)
                 .unwrap_or(0);
-            let lambda_1_i = ((delta_2_i as u128) * (self.lambda_1 as u128))
+        let lambda_1_i = ((delta_1_i as u128) * (self.lambda_2_1 as u128))
+            .checked_div(self.delta_1 as u128)
+            .unwrap_or(0)
+            + ((delta_2_i as u128) * (self.lambda_1_2 as u128))
                 .checked_div(self.delta_2 as u128)
                 .unwrap_or(0);
 
-            (lambda_1_i as u64, lambda_2_i as u64)
-        } else {
-            // The swap failed, so the pro rata share is a share of the input amount of
-            // the same token type. But this is exactly the delta_j_i.
-            (delta_1_i, delta_2_i)
-        }
+        (lambda_1_i as u64, lambda_2_i as u64)
```

Comment:
This code was written before we had the U128x128, and I think it's probably not going to correctly handle trades with 128-bit inputs (actually, the BSOD can't even currently model those, since it uses `u64` rather than `Amount`s)...

Maybe we should take the opportunity to change the BSOD's data types from `u64` to `Amount`, and then use the `U128x128` to implement the clean versions of the formulas rather than doing manual casting and rounding.